clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -4536,23 +4548,162 @@ void SYCLIntegrationFooter::emitSpecIDName(raw_ostream &O, const VarDecl *VD) {
   O << "";
 }
 
+template <typename Before, typename After>
+static void PrintNSHelper(Before B, After A, raw_ostream &O,
+                          const DeclContext *DC) {
+  if (DC->isTranslationUnit())
+    return;
+
+  const auto *CurDecl = cast<Decl>(DC);
+  // Ensure we are in the canonical version, so that we know we have the 'full'
+  // name of the thing.
+  CurDecl = CurDecl->getCanonicalDecl();
+
+  // We are intentionally skipping linkage decls and record decls.  Namespaces
+  // can appear in a linkage decl, but not a record decl, so we don't have to
+  // worry about the names getting messed up from that.  We handle record-decls
+  // later when printing the name of the thing.
+  if (const auto *NS = dyn_cast<NamespaceDecl>(CurDecl))
+    B(O, NS);
+
+  if (CurDecl->getDeclContext())
+    PrintNSHelper(B, A, O, CurDecl->getDeclContext());
+
+  if (const auto *NS = dyn_cast<NamespaceDecl>(CurDecl))
+    A(O, NS);
+}
+
+static void PrintNamespaces(raw_ostream &O, const DeclContext *DC) {
+  PrintNSHelper([](raw_ostream &O, const NamespaceDecl *NS) {},
+                [](raw_ostream &O, const NamespaceDecl *NS) {
+                  if (NS->isInline())
+                    O << "inline ";
+                  O << "namespace ";
+                  if (!NS->isAnonymousNamespace())
+                    O << NS->getName() << " ";
+                  O << "{\n";
+                },
+                O, DC);
+}
+
+static void PrintNSClosingBraces(raw_ostream &O, const DeclContext *DC) {
+  PrintNSHelper(
+      [](raw_ostream &O, const NamespaceDecl *NS) {
+        O << "} // ";
+        if (NS->isInline())
+          O << "inline ";
+
+        O << "namespace ";
+        if (!NS->isAnonymousNamespace())
+          O << NS->getName();
+
+        O << '\n';
+      },
+      [](raw_ostream &O, const NamespaceDecl *NS) {}, O, DC);
+}
+
+static std::string EmitSpecIdShim(raw_ostream &O, unsigned &ShimCounter,
+                                  const std::string &LastShim,
+                                  const NamespaceDecl *AnonNS) {
+  std::string NewShimName =
+      "__sycl_detail::__spec_id_shim_" + std::to_string(ShimCounter) + "()";
+  // Print opening-namespace
+  PrintNamespaces(O, Decl::castToDeclContext(AnonNS));
+  O << "namespace __sycl_detail {\n";
+  O << "static constexpr decltype(" << LastShim << ") &__spec_id_shim_"
+    << ShimCounter << "() {\n";
+  O << "  return " << LastShim << ";\n";
+  O << "}\n";
+  O << "} // namespace __sycl_detail \n";
+  PrintNSClosingBraces(O, Decl::castToDeclContext(AnonNS));
+
+  ++ShimCounter;
+  return std::move(NewShimName);
+}
+
+// Emit the list of shims required for a DeclContext, calls itself recursively.
+static std::string EmitSpecIdShims(raw_ostream &O, unsigned &ShimCounter,
+                                   const DeclContext *DC,
+                                   std::string NameForLastShim) {
```

Comment:
Can NameForLastShim be a reference?

Reply:
The intent was for it to be a 'local variable' here so that it could modify it without affecting the parent.  That said, it seems that it only does tail-call, so I believe it can.  We just need to make sure with the change that we never allow anything that modifies this variable to be below the recursion.